File: hello-go/src/client/util/verification_module.go
package util

import (
	"fmt"
	"strings"

	"github.com/feuyeux/hello-mmap/hello-go/src/logger"
)

type VerificationResult struct {
	Passed             bool
	OriginalSize       int64
	DownloadedSize     int64
	OriginalChecksum   string
	DownloadedChecksum string
}

func Verify(originalPath string, downloadedPath string) (*VerificationResult, error) {
	logger.Info(fmt.Sprintf("Original file: %s", originalPath))
	logger.Info(fmt.Sprintf("Downloaded file: %s", downloadedPath))

	// Get file sizes
	originalSize, err := GetFileSize(originalPath)
	if err != nil {
		return nil, fmt.Errorf("failed to get original file size: %w", err)
	}

	downloadedSize, err := GetFileSize(downloadedPath)
	if err != nil {
		return nil, fmt.Errorf("failed to get downloaded file size: %w", err)
	}

	logger.Info(fmt.Sprintf("Original size: %d bytes", originalSize))
	logger.Info(fmt.Sprintf("Downloaded size: %d bytes", downloadedSize))

	// Compute checksums
	originalChecksum, err := ComputeSHA256(originalPath)
	if err != nil {
		return nil, fmt.Errorf("failed to compute original checksum: %w", err)
	}

	downloadedChecksum, err := ComputeSHA256(downloadedPath)
	if err != nil {
		return nil, fmt.Errorf("failed to compute downloaded checksum: %w", err)
	}

	logger.Info(fmt.Sprintf("Original checksum (SHA-256): %s", originalChecksum))
	logger.Info(fmt.Sprintf("Downloaded checksum (SHA-256): %s", downloadedChecksum))

	// Compare
	passed := originalSize == downloadedSize &&
		strings.EqualFold(originalChecksum, downloadedChecksum)

	return &VerificationResult{
		Passed:             passed,
		OriginalSize:       originalSize,
		DownloadedSize:     downloadedSize,
		OriginalChecksum:   originalChecksum,
		DownloadedChecksum: downloadedChecksum,
	}, nil
}

File: hello-rust/src/client/verification_module.rs
use super::file_manager;
use crate::logger;
use anyhow::Result;

pub struct VerificationResult {
    pub passed: bool,
    pub original_size: u64,
    pub downloaded_size: u64,
    pub original_checksum: String,
    pub downloaded_checksum: String,
}

pub async fn verify(original_path: &str, downloaded_path: &str) -> Result<VerificationResult> {
    logger::log_info(&format!("Original file: {}", original_path));
    logger::log_info(&format!("Downloaded file: {}", downloaded_path));

    // Get file sizes
    let original_size = file_manager::get_file_size(original_path)?;
    let downloaded_size = file_manager::get_file_size(downloaded_path)?;

    logger::log_info(&format!("Original size: {} bytes", original_size));
    logger::log_info(&format!("Downloaded size: {} bytes", downloaded_size));

    // Compute checksums
    let original_checksum = file_manager::compute_sha256(original_path).await?;
    let downloaded_checksum = file_manager::compute_sha256(downloaded_path).await?;

    logger::log_info(&format!(
        "Original checksum (SHA-256): {}",
        original_checksum
    ));
    logger::log_info(&format!(
        "Downloaded checksum (SHA-256): {}",
        downloaded_checksum
    ));

    // Compare
    let passed = original_size == downloaded_size
        && original_checksum.to_lowercase() == downloaded_checksum.to_lowercase();

    Ok(VerificationResult {
        passed,
        original_size,
        downloaded_size,
        original_checksum,
        downloaded_checksum,
    })
}
